clock: provide k_cycle_get_64() when hardware supports a 64-bit counter

cfriedt opened the pull request:
This change adds `k_cycle_get_64()` on platforms that support a 64-bit cycle counter.

The interface functions `arch_k_cycle_get_64()` and `sys_clock_cycle_get_64()` are also introduced.

Fixes #39934

cfriedt committed:
clock: add k_cycle_get_64

This change adds `k_cycle_get_64()` on platforms that
support a 64-bit cycle counter.

The interface functions `arch_k_cycle_get_64()` and
`sys_clock_cycle_get_64()` are also introduced.

Fixes #39934

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt committed:
tests: clock: tests for k_cycle_get_64

This change adds tests for 64-bit cycle counters.

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

cfriedt commented on drivers/timer/altera_avalon_timer_hal.c:
```diff
@@ -81,6 +81,11 @@ uint32_t sys_clock_cycle_get_32(void)
 
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@nashif - would be great to get your input here. It seems there is no 64-bit timer register on this platform.

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

cfriedt commented on drivers/timer/apic_timer.c:
```diff
@@ -211,6 +211,11 @@ uint32_t sys_clock_cycle_get_32(void)
 
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@nashif - would be great to get your input here. It seems there is no 64-bit timer register on this platform.

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

andyross replied:
Actually on all modern hardware the APIC LVT timer is slaved to the TSC (with an configurable shift), so it's possible to synthesize a 64 bit value.  That's probably not worth it though.

I'm definitely not a fan of adding runtime complexity to a timer driver just to emulate a rarely-needed feature.  The overwhelmingly common use of cycle timing is to compute tiny delays, after all.  It's not worth taking a performance or code size penalty on hardware just to support oddball use cases.

I guess my preference here would be to make the 64 bit cycle API an optional thing, gated by a kconfig.  You can use the API if your timer hardware supports it, and not if it doesn't.  That's more or less the way it works for ticks (c.f. CONFIG_TIMEOUT_64BIT), though the concern there is code size and not hardware capability as the kernel is willing to do the wrap logic internally.

cfriedt commented on drivers/timer/ite_it8xxx2_timer.c:
```diff
@@ -229,6 +229,11 @@ uint32_t sys_clock_cycle_get_32(void)
 
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@Dino-Li  - would be great to get your input here. It seems there is no 64-bit timer register on this platform.

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

Dino-Li replied:
I created PR here: #39981

cfriedt commented on drivers/timer/leon_gptimer.c:
```diff
@@ -95,6 +95,11 @@ uint32_t sys_clock_cycle_get_32(void)
 
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@martin-aberg  - would be great to get your input here. It seems there is no 64-bit timer register on this platform.

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

cfriedt commented on drivers/timer/mchp_xec_rtos_timer.c:
```diff
@@ -353,6 +353,11 @@ uint32_t sys_clock_cycle_get_32(void)
 
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@scottwcpg   - would be great to get your input here. It seems there is no 64-bit timer register on this platform.

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

cfriedt commented on drivers/timer/mcux_lptmr_timer.c:
```diff
@@ -81,6 +81,11 @@ uint32_t sys_clock_cycle_get_32(void)
 
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@henrikbrixandersen   - would be great to get your input here. It seems there is no 64-bit timer register on this platform.

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

cfriedt commented on boards/arm/qemu_cortex_m0/nrf_timer_timer.c:
```diff
@@ -255,3 +255,8 @@ uint32_t sys_clock_cycle_get_32(void)
+
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@ioannisg - would be great to get your input here. It seems there is no 64-bit timer register on this platform.

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

cfriedt commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -409,3 +409,8 @@ uint32_t sys_clock_cycle_get_32(void)
+
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@nordic-krch - would be great to get your input here. It seems there is no 64-bit timer register on this platform.

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

nordic-krch replied:
We have #39099 were we will add support to 64 bit timestamps. It will be safe to read from any context (including ZLI). However, it is not yet finished.

cfriedt commented on drivers/timer/rcar_cmt_timer.c:
```diff
@@ -150,3 +150,8 @@ uint32_t sys_clock_cycle_get_32(void)
+
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@julien-massot - would be great to get your input here. It seems there is no 64-bit timer register on this platform.

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

julien-massot replied:
Hi @cfriedt, what I can do is to connect the overflow interrupt and increments a 32 bit top half.
But one rule about the sys_clock_cycle_get_32 is that it should work even with irq disabled, to allow k_busy_wait
to work with irq locked or within high priority ISR.  

So in case we increments the top half in the overflow interrupts, sys_clock_cycle_get_32 will be reliable
with irq locked and sys_clock_cycle_get_64 will not.

cfriedt replied:
Interesting - @stephanosio also suggested another solution for when overflow IRQs are not available is to use the 64-bit system tick counter + some arithmetic for the upper 32-bits.

cfriedt commented on drivers/timer/rv32m1_lptmr_timer.c:
```diff
@@ -136,6 +136,11 @@ uint32_t sys_clock_cycle_get_32(void)
 
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@mbolivar-nordic - would be great to get your input here. It seems there is no 64-bit timer register on this platform (i.e. the lptimer).

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

cfriedt commented on drivers/timer/sam0_rtc_timer.c:
```diff
@@ -315,3 +315,8 @@ uint32_t sys_clock_cycle_get_32(void)
+
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@bendiscz - would be great to get your input here. It seems there is no 64-bit timer register on this platform (i.e. the lptimer).

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

cfriedt commented on drivers/timer/stm32_lptim_timer.c:
```diff
@@ -374,3 +374,8 @@ uint32_t sys_clock_cycle_get_32(void)
+
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@erwango  - would be great to get your input here. It seems there is no 64-bit timer register on this platform (i.e. the lptimer).

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

erwango replied:
@FRASTM, can you have a check?

cfriedt commented on drivers/timer/xlnx_psttc_timer.c:
```diff
@@ -195,3 +195,8 @@ uint32_t sys_clock_cycle_get_32(void)
+
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@stephanosio - would be great to get your input here. It seems there is no 64-bit timer register on this platform (i.e. the lptimer).

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

cfriedt commented on drivers/timer/xtensa_sys_timer.c:
```diff
@@ -115,6 +115,11 @@ uint32_t sys_clock_cycle_get_32(void)
 
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
```
@andyross - would be great to get your input here. It seems there is no 64-bit timer register on this platform (i.e. the lptimer).

For software emulation, I would like to increment a 32-bit "top half" each time a 32-bit cycle counter overflows (assuming a 1:1 ratio), shift and then OR it with the regular 32-bit value.

In this driver, where is the best location to increment the 32-bit "top half"?

cfriedt commented on tests/kernel/common/src/clock.c:
```diff
@@ -160,6 +160,35 @@ void test_clock_cycle(void)
 
+/**
+ * @brief Test clock cycle 64 functionality
+ */
```
This test really only checks to ensure that the 64-bit cycle counter grows proportionately to the 32-bit cycle counter on each platform.

I might add a sample application that periodically checks to ensure that the 64-bit timer has received updates to the upper 32-bits based on `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC`.

stephanosio reviewed (commented):
> For platforms / drivers where there is not a 64-bit register (or a pair of lo / hi 32-bit registers), the approach I would like to take is to increment a 32-bit value when we have detected that the 32-bit counter has overflowed.

The kernel already keeps track of the uptime ticks as 64-bit integer, so `k_uptime_ticks` effectively achieves the same thing as what is described here.

My concern is:
1. Given that there are not that many hardware timers out there (especially for MCUs) that support 64-bit count register, does it make sense to define `sys_clock_cycle_get_64`?
2. Should a timer hardware driver be responsible for emulating `sys_clock_cycle_get_64` when the hardware itself does not provide a 64-bit count register? IMO, the driver should just return something like `ENOTSUP` and the kernel should be responsible for such emulations/abstractions.
3. Is `sys_clock_cycle_get_64` really all that useful when we have `k_uptime_ticks` that returns 64-bit uptime tick count?

cfriedt commented:
> > For platforms / drivers where there is not a 64-bit register (or a pair of lo / hi 32-bit registers), the approach I would like to take is to increment a 32-bit value when we have detected that the 32-bit counter has overflowed.
> 
> The kernel already keeps track of the uptime ticks as 64-bit integer, so `k_uptime_ticks` effectively achieves the same thing as what is described here.

Interesting. On platforms without a 64-bit counter (and in particular when a 32-bit overflow interrupt is not really available), it might make sense to simply perform some arithmetic on the 64-bit tick counter to get the "top half". E.g. 

```
// unshifted
top_half = ((k_uptime_ticks() * CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC) / SYS_CLOCK_TICKS_PER_SEC) & ~((uint64_t)UINT32_T_MAX);
```

> My concern is:
> 
> 1. Given that there are not that many hardware timers out there (especially for MCUs) that support 64-bit count register, does it make sense to define `sys_clock_cycle_get_64`?

I think it does make sense, for consistency. 

> 2. Should a timer hardware driver be responsible for emulating `sys_clock_cycle_get_64` when the hardware itself does not provide a 64-bit count register? IMO, the driver should just return something like `ENOTSUP` and the kernel should be responsible for such emulations/abstractions.

That's one option. Given that most 64-bit hardware timers will not overflow for thousands of years, a negative return value from `k_cycle_get()` could mean a failure (e.g. `-ENOTSUP`).

> 3. Is `sys_clock_cycle_get_64` really all that useful when we have `k_uptime_ticks` that returns 64-bit uptime tick count?

Yes, it absolutely is useful when there are machines that could make use of that information for e.g. profiling.

nordic-krch commented:
> Given that there are not that many hardware timers out there (especially for MCUs) that support 64-bit count register, does it make sense to define sys_clock_cycle_get_64?

From logging time perspective, it is best to have fast time stamping mechanism and hw reading may be faster than calculated sys tick. Additionally, if `sys_clock_cycle_get_64` would return negative error when not supported it could be used to setup at runtime timestamping source.

julien-massot commented on drivers/timer/esp32c3_sys_timer.c:
```diff
@@ -134,3 +134,15 @@ uint32_t sys_clock_cycle_get_32(void)
+	ret |= systimer_ll_get_counter_value_low(SYSTIMER_COUNTER_1);
+
+	k_spin_unlock(&lock, key);
+
```
Common scheme is to read the high value again after reading the lower value.
see ./include/arch/arm/aarch32/cortex_a_r/timer.h
```
	uint32_t lower;
	uint32_t upper, upper_saved;

	/* To get the value from the Global Timer Counter register proceed
	 * as follows:
	 * 1. Read the upper 32-bit timer counter register.
	 * 2. Read the lower 32-bit timer counter register.
	 * 3. Read the upper 32-bit timer counter register again. If the value
	 * is different to the 32-bit upper value read previously,
	 * go back to step 2.
	 * Otherwise the 64-bit timer counter value is correct.
	 */
	upper = sys_read32(TIMER_REG_GET(TIMER_CNT_UPPER));
	do {
		upper_saved = upper;
		lower = sys_read32(TIMER_REG_GET(TIMER_CNT_LOWER));
		upper = sys_read32(TIMER_REG_GET(TIMER_CNT_UPPER));
	} while (upper != upper_saved);

	return ((uint64_t)upper) << 32 | lower;

cfriedt committed:
samples: kernel: add cycle64 sample

The cycle64 sample is intended to complement
`test_clock_cycle_64()` in `tests/kernel/common`.

The sample demonstrates the upper 32-bits of the 64-bit cycle
counter incrementing when the bottom 32-bits roll over from
`UINT32_MAX` to 0.

If the upper 32-bits of the 64-bit cycle counter does not
increment, then an error message is printed.

```
west build -p auto -b qemu_cortex_a53 -t run \
	samples/kernel/cycle64
...
*** Booting Zephyr OS build v2.7.99-1124-gd7ba4e394832  ***
wrap-around should occur in 68s
[ddd:hh:mm:ss.0ms]
[000:00:00:00.020]: c64: 0000000000174258
[000:00:01:08.760]: c64: 000000010027f8bb
[000:00:02:17.490]: c64: 0000000200348c85
```

Signed-off-by: Christopher Friedt <chrisfriedt@gmail.com>

ruuddw commented on include/arch/arc/v2/misc.h:
```diff
@@ -27,6 +27,13 @@ static inline uint32_t arch_k_cycle_get_32(void)
+
+static inline uint64_t arch_k_cycle_get_64(void)
+{
+	return sys_clock_cycle_get_64();
```
Depending on the specific hardware configuration of a specific ARC CPU, there could be a 64-bit counter available.
We can add support for this later.
One question: if both 32-bit and 64-bit cycle counters are provided, are there any constraints or expectations on their relation? E.g. should the 32-bit one always be the least significant bits of the 64-bit one? Same frequency?

cfriedt replied:
Technically speaking, the lower 32 bits do not need to be the same, but they should have the same frequency

brockus-zephyr replied:
I think it is best for the two to be synchronized and the lower 32 bits of the 64 bit timer should always be the same as the 32 bit timer.

andyross replied:
Pedantry: there's no way to express that requirement in a way that is actually implementable.  This isn't a synchronization problem, you can't "lock" against the passage of time.   At best we should just specify that the two APIs should ideally be backed by the same clock, or at worst have no practically detectable offset nor rate skew.

brockus-zephyr replied:
I did not intend this to be stopping the clock between different reads... I am saying if you do the following that it will be as expected from a single timer.

1. first_read = read 64 bit timer
2. second_read = read 32 bit timer
3. third_read = read 64 bit timer

(first_read & 0xFFFFFFFF) <= second_read <= (third_read & 0xFFFFFFFF)

ruuddw replied:
Agreed - my main point was indeed that the requirement or expected behavior is somewhere specified and documented. I guess there are pro's and con's for both approaches, but since I would not expect both to be used and combined (comparing 32-bit one with 64 bit one or so), I'd say only have the same frequency.

carlocaione reviewed (approved):
LGMT for the ARM64 part

stephanosio commented on tests/kernel/common/src/clock.c:
```diff
@@ -160,6 +160,39 @@ void test_clock_cycle(void)
+	zassert_true(d64 < (d32 << 1),
+		"k_cycle_get() (64-bit): d64: %" PRIu64 " >= 2 * d32: %u",
+		d64, (d32 << 1));
+
```
```
#else
	ztest_test_skip();
```

stephanosio commented:
> > 3. Is `sys_clock_cycle_get_64` really all that useful when we have `k_uptime_ticks` that returns 64-bit uptime tick count?
> 
> If the hardware is there and it's supported, it really does make sense to take advantage of the numerical accuracy, because ticks are often not as accurate.

The "ticks" returned by `k_uptime_ticks` are the system timer ticks.

https://github.com/zephyrproject-rtos/zephyr/blob/339a6bdafbe558e3ad0ea57336e2ead62e2237ac/kernel/timeout.c#L288-L291
https://github.com/zephyrproject-rtos/zephyr/blob/339a6bdafbe558e3ad0ea57336e2ead62e2237ac/kernel/timeout.c#L269-L277

Unless you have a buggy system timer driver that returns incorrect value from `sys_clock_elapsed`, the value of `k_uptime_ticks` and the proposed `sys_clock_cycle_get_64` here should be the same (in theory).

evgeniy-paltsev commented on kernel/Kconfig:
```diff
@@ -627,6 +627,12 @@ config SYS_CLOCK_EXISTS
 	  this is disabled.  Obviously timeout-related APIs will not
 	  work.
 
+config CYCLE_64BIT
```
Based on the usage of this Kconfig symbol we should name with `HAS`. Probably we can use something like `CPU_HAS_CYCLE_64BIT`

stephanosio replied:
> `CPU_HAS_CYCLE_64BIT`

Not sure about the `CPU_` part, because it does not necessarily have to do with the **CPU** (especially for most of the system timer drivers that use a separate timer peripheral that is independent of the CPU).

Maybe `TIMER_HAS_CYCLE_64BIT` in `drivers/timer/Kconfig` that is selected by each timer driver (similar to `TICKLESS_CAPABLE`), and a separate config that enables 64-bit cycle support in the kernel (e.g. `CYCLE_64BIT`) that depends on `TIMER_HAS_CYCLE_64BIT`.

stephanosio reviewed (commented):
Looks ok overall. Some minor comments.

jhedberg commented on include/kernel.h:
```diff
@@ -1639,6 +1639,21 @@ static inline uint32_t k_cycle_get_32(void)
 	return arch_k_cycle_get_32();
 }
 
+#if defined(CONFIG_CYCLE_64BIT)
```
I think this kind of #ifdefs in public header files are generally discouraged, as then the documentation doesn't get generated unless you add an exception to `doc/zephyr.doxyfile.in`. Just omitting the #ifdef should give an equally understandable build error if an app tries to use this API on a platform that doesn't support it.

cfriedt commented:
> > > 1. Is `sys_clock_cycle_get_64` really all that useful when we have `k_uptime_ticks` that returns 64-bit uptime tick count?
> > 
> > 
> > If the hardware is there and it's supported, it really does make sense to take advantage of the numerical accuracy, because ticks are often not as accurate.
> 
> The "ticks" returned by `k_uptime_ticks` are the system timer ticks.
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/339a6bdafbe558e3ad0ea57336e2ead62e2237ac/kernel/timeout.c#L288-L291
> 
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/339a6bdafbe558e3ad0ea57336e2ead62e2237ac/kernel/timeout.c#L269-L277
> 
> Unless you have a buggy system timer driver that returns incorrect value from `sys_clock_elapsed`, the value of `k_uptime_ticks` and the proposed `sys_clock_cycle_get_64` here should be the same (in theory).

I have seen a number of drivers that return 0 for `elapsed()`. But really why differentiate "ticks" and "cycles" if they are the same? From what I understand, they are different concepts, hence
```
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
CONFIG_SYS_CLOCK_TICKS_PER_SEC
```

> It does seem that the value returned by `sys_clock_cycle_get` can be independent of the timer "ticks," although almost all system timer drivers we currently have seem to return the timer tick value.

Perhaps it's a good time to unify that concept.

stephanosio commented:
> I have seen a number of drivers that return 0 for `elapsed()`. But really why differentiate "ticks" and "cycles" if they are the same? From what I understand, they are different concepts, hence
> 
> ```
> CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
> CONFIG_SYS_CLOCK_TICKS_PER_SEC
> ```
> 
> > It does seem that the value returned by `sys_clock_cycle_get` can be independent of the timer "ticks," although almost all system timer drivers we currently have seem to return the timer tick value.
> 
> Perhaps it's a good time to unify that concept.

We currently have many drivers that define something called `CYCLES_PER_TICK`, which directly correlates "ticks" to "cycles"; but, according to the documentation, that does not necessarily have to be the case (e.g. "cycle" counter/timer can be a completely separate counter/timer instance from the "tick" counter/timer, that is running in a separate clock domain).

IMO, in that case, there should be two separate drivers: low frequency timer driver for "ticks" (i.e. for tickless kernel/wakeup) and high frequency timer driver for things like nanosleep.

Ideally, the whole thing should be refactored to use some form of generalised timer driver interface that can be mapped to each timer class via the `chosen` nodes in the devicetree, as we do with the rest of the drivers; but, that is something for another day.

Perhaps, someone who architected this can give us a better insight as to why it was done this way and if there is any reason not to do what I suggested above; though, my guess is that it was likely done this way because this "system timer driver" scheme predates the introduction of devicetree.

cfriedt commented:
About to update with all of the changes requested, just waiting for twister..

andyross commented:
@stephanosio That's right.  Cycles and ticks are conceptually different counters running at different rates and potentially skewed relative to each other (i.e. on the default nRF5 configuration, you will see the two clocks drift because one is a crystal and the other a counted software loop on a CPU clocked by the unsynchronized internal oscillator).

It's just the case that on most platforms you implement "cycles" as the low level counter of the same device used to implement "ticks".

andyross commented:
Oh, and indeed the timer driver has been just a plain C API since time immemorial.  In practice it ends up being very tightly coupled to arch/soc/board level code, though.  While there are certainly some platforms that could benefit from a devicetree/pluggable setup, I don't see a lot of real value there.  Where we have had the opportunity to select timers (HPET, I'm looking at you) it's generally caused more headache than help.
